### PR TITLE
Add Crypto dataset from coingecko

### PR DESCRIPTION
**Pull request by b4thesunrise**
## Description

Add collector in scripts to add crypto data
## Motivation and Context

https://github.com/microsoft/qlib/issues/716

solve the lack of crypto data, which is a need for many people(this need has been raised in gitter also)
## How Has This Been Tested?
- [] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [x] If you are adding a new feature, test on your own test scripts.
ran the scripts in readme successfully

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: scripts/data_collector/crypto/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+## Collector Data
+
+
+### Crypto Data
+
+#### 1d from Coingecko
+
+```bash
+
+# download from eastmoney.com
```
Comment (zhupr) [inline scripts/data_collector/crypto/README.md:20]: `download from https://api.coingecko.com/api/v3/`

File: scripts/data_collector/crypto/collector.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    def _get_calendar_list(self):
+        return None
+
+
+class CryptoNormalizeCN1d(CryptoNormalizeCN, CryptoNormalize1d):
+    pass
+
+
+class Run(BaseRun):
+    def __init__(self, source_dir=None, normalize_dir=None, max_workers=1, interval="1d", region=REGION_CN):
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:208]: Remove the `region` parameter

**Comment by zhupr** [comment]
@b4thesunrise 
Thanks for your contribution.
Crypto is `region` free and can be removed `CryptoXXXCN`

**Comment by b4thesunrise** [comment]
> @b4thesunrise Thanks for your contribution. Crypto is `region` free and can be removed `CryptoXXXCN`

thanks for your help! will remove it in later commit.

**Comment by b4thesunrise** [comment]
> commented

have already implement the change, please approve for check :)

File: scripts/data_collector/crypto/collector.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+            )
+
+        if interval == self.INTERVAL_1d:
+            _result = _get_simple(start_datetime, end_datetime)
+        else:
+            raise ValueError(f"cannot support {interval}")
+        return _result
+
+
+class CryptoCollectorCN(CryptoCollector, ABC):
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:140]: `CryptoCollectorCN`  renamed to `CryptoCollector1d`

File: scripts/data_collector/crypto/collector.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    def normalize_symbol(self, symbol):
+        return symbol
+
+    @property
+    def _timezone(self):
+        return "Asia/Shanghai"
+
+
+class CryptoCollectorCN1d(CryptoCollectorCN):
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:155]: remove `CryptoCollectorCN1d`

File: scripts/data_collector/crypto/collector.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+        df.index.names = [date_field_name]
+        return df.reset_index()
+
+    def normalize(self, df: pd.DataFrame) -> pd.DataFrame:
+        df = self.normalize_crypto(df, self._calendar_list, self._date_field_name, self._symbol_field_name)
+        return df
+
+
+class CryptoNormalize1d(CryptoNormalize):
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:194]:
```python
class CryptoNormalize1d(CryptoNormalize):
    def _get_calendar_list(self):
        return None
```
remove `CryptoNormalizeCN` and `CryptoNormalizeCN1d`

File: scripts/data_collector/crypto/collector.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+            Directory for normalize data, default "Path(__file__).parent/normalize"
+        max_workers: int
+            Concurrent number, default is 4
+        interval: str
+            freq, value from [1min, 1d], default 1d
+        region: str
+            region, value from ["CN"], default "CN"
+        """
+        super().__init__(source_dir, normalize_dir, max_workers, interval)
+        self.region = region
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:225]: remove `self.region = region`

File: scripts/data_collector/crypto/collector.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+    @property
+    def collector_class_name(self):
+        return f"CryptoCollector{self.region.upper()}{self.interval}"
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:229]: `return f"CryptoCollector{self.region.upper()}{self.interval}"` to `return f"CryptoCollector{self.interval}"`

File: scripts/data_collector/crypto/collector.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+    @property
+    def normalize_class_name(self):
+        return f"CryptoNormalize{self.region.upper()}{self.interval}"
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:233]: as above

File: scripts/data_collector/crypto/collector.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            end datetime, default ``pd.Timestamp(datetime.datetime.now() + pd.Timedelta(days=1))``
+        check_data_length: int # if this param useful?
+            check data length, if not None and greater than 0, each symbol will be considered complete if its data length is greater than or equal to this value, otherwise it will be fetched again, the maximum number of fetches being (max_collector_count). By default None.
+        limit_nums: int
+            using for debug, by default None
+
+        Examples
+        ---------
+            # get daily data
+            $ python collector.py download_data --source_dir ~/.qlid/crypto_data/source/cn_1d --region CN --start 2015-01-01 --end 2021-12-01 --delay 1 --interval 1d
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:271]: remove `--region CN`

File: scripts/data_collector/crypto/collector.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        Parameters
+        ----------
+        date_field_name: str
+            date field name, default date
+        symbol_field_name: str
+            symbol field name, default symbol
+
+        Examples
+        ---------
+            $ python collector.py normalize_data --source_dir ~/.qlib/crypto_data/source/cn_1d --normalize_dir ~/.qlib/crypto_data/source/cn_1d_nor --region CN --interval 1d --date_field_name date
```
Comment (zhupr) [inline scripts/data_collector/crypto/collector.py:288]: as above

File: scripts/data_collector/crypto/README.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+
+### Crypto Data
+
+#### 1d from Coingecko
+
+```bash
+
+# download from https://api.coingecko.com/api/v3/
+python collector.py download_data --source_dir ~/.qlib/crypto_data/source/cn_1d --start 2015-01-01 --end 2021-11-30 --delay 1 --interval 1d
```
Comment (zhupr) [inline scripts/data_collector/crypto/README.md:21]: remove region-related information from the day's documentation, such as `cn`

**Comment by zhupr** [comment]
> > commented
> 
> have already implement the change, please give it a review :)

Hi, currently `CryptoCollector` can only be used in `Data retrieval`, does not support `Backtest`, if users need to support `backtest`, they need to meet "Data-dependent component: Backtest", reference: https://github.com/microsoft/qlib/tree/main/scripts/data_collector#description-of-dataset

**Comment by you-n-g** [comment]
@b4thesunrise 
I think the following 2 choices could make it clearer for users.
1. Describe the content and limitation (e.g. it don't support backtesting) of the dataset in the documentation.
2. Implement a complete version of the dataset to make all the data dependant components runnable

You can choose either of them as you wish.
Thanks.

**Comment by b4thesunrise** [comment]
> > > commented
> > 
> > 
> > have already implement the change, please give it a review :)
> 
> Hi, currently `CryptoCollector` can only be used in `Data retrieval`, does not support `Backtest`, if users need to support `backtest`, they need to meet "Data-dependent component: Backtest", reference: https://github.com/microsoft/qlib/tree/main/scripts/data_collector#description-of-dataset

Hi, so to support backtest, the data need to support all the features of  $close/$open/$low/$high/$volume/$change/$factor, don't know if I understand it correctly? If true, the currently data source do not support OHLC and I will choose to describe the limitation in the file.

**Comment by you-n-g** [comment]
@b4thesunrise Yes. You understand it correctly.
Please add some docs to describe the limitation of the data source.

**Comment by b4thesunrise** [comment]
> @b4thesunrise Yes. You understand it correctly. Please add some docs to describe the limitation of the data source.

Got it, so busy for these two weeks, plan to update the docs and new code version at 12.30.

**Comment by b4thesunrise** [comment]
> @b4thesunrise Yes. You understand it correctly. Please add some docs to describe the limitation of the data source.
Have already clean the cn symbols and change the documentation, please give a review:)

**Comment by you-n-g** [comment]
It looks great!
Thanks so much!

**Pull request merged**